src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+    where
+        R: io::BufRead;
+
+    /// The primary method.
+    ///
+    /// It receives a reader whose header has already been consumed (see above), as
+    /// well as a mutable reference to an `Occurences` instance (just a hashamp that
+    /// works across multiple threads). Implementators should parse the stack data
+    /// contained in the reader and write output to the provided `Occurrences` map.
+    ///
+    /// So that this method may be called multiple times, all internal
+    /// state contained in `self` (e.g. any stack buffers or the like) **must** be
+    /// reset before this method returns. Also, this method may **not** use threads.
+    fn collapse_single_threaded<R>(
+        &mut self,
+        reader: R,
+        occurrences: &mut Occurrences,
+    ) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// This method receives a line from the input data (including the trailing newline byte)
+    /// and should return `true` if the line represents the end of a stack; `false` otherwise.
+    fn is_end_of_stack(&self, line: &[u8]) -> bool;
+
+    /// This method should return whether the implementation corresponds with
+    /// the given input string, i.e. if the input data matches the collapser.
+    ///
+    /// - `None` means "not sure -- need more input"
+    /// - `Some(true)` means "yes, this implementation should work with this string"
+    /// - `Some(false)` means "no, this implementation definitely won't work"
+    fn is_applicable_(&mut self, input: &str) -> Option<bool>;
+
+    /// This method should return the number of stacks per job to send to the threadpool.
+    fn nstacks_per_job(&self) -> usize;
+
+    /// This method should set the number of stacks per job to send to the threadpool.
+    fn set_nstacks_per_job(&mut self, n: usize);
+
+    /// This method should return the number of threads to use.
+    fn nthreads(&self) -> usize;
+
+    /// This method should set the number of threads to use.
+    fn set_nthreads_(&mut self, n: usize);
+
+    // *********************************************************** //
+    // ******************** PROVIDED METHODS ********************* //
+    // *********************************************************** //
+
+    fn collapse_<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: io::BufRead,
+        W: io::Write,
+    {
+        let mut occurrences = Occurrences::new(self.nthreads());
+
+        // Consume the header, if any, and do any other pre-processing
+        // that needs to occur.
+        self.pre_process(&mut reader, &mut occurrences)?;
+
+        // Do collapsing.
+        if occurrences.is_concurrent() {
+            self.collapse_multi_threaded(reader, &mut occurrences)?;
+        } else {
+            self.collapse_single_threaded(reader, &mut occurrences)?;
+        }
+
+        // Write results.
+        occurrences.write_and_clear(writer)
+    }
+
+    fn collapse_multi_threaded<R>(
```

Comment:
Callers should probably never call this method directly, only ever `collapse`. We should call that out in docs. Alternatively, I wonder whether this whole multi-threading business should be represented as a `struct` that is generic over a collapser that implements `CollapsePrivate`. Rather than putting all this implementation in the trait itself.

Reply:
They can't. `CollapsePrivate` is in a private module; so there is no way for users of the library to use any of it's methods.